### PR TITLE
Add Enum constants for trait kinds, validators and default values

### PR DESCRIPTION
**Pull request by corranwebster**
This is an attempt to remove as many magic numbers as possible from the traits source code, plus use new Python 3 features to do it.

Possibly should also add some C constants for validator types to ctraits.c.

### CODE REVIEW

File: traits/api.py [hunk 1]
```diff
@@ -230,3 +229,8 @@
         AbstractViewElement.register(ViewElement)
 except ImportError:
     pass
+
+# Backward compatibility for comparison mode constants.
```
Comment (mdickinson) [inline traits/api.py]: Is this just for backward compatibility? E.g., should we recommend that people use `ComparisonMode.no_compare` in preference to `NO_COMPARE` going forward? If so, we should update the documentation accordingly.

File: traits/constants.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+
+class TraitKind(IntEnum):
+    """ These determine the getters and setters used by the cTrait instance. """
+
+    #: A standard trait (validates and notifies).
+    trait = 0
+
+    #: A standard Python attribute (no validation or notification).
+    #: This sets valus into the object dict, so is not suitable for use with
```
Comment (mdickinson) [inline traits/constants.py]: typo: values

File: traits/constants.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    disallow = 5
+
+    #: A trait that can be assigned once, and after that only read (no
+    #: validation or notification).
+    read_only = 6
+
+    #: A trait whose value is constant (can only be read).
+    constant = 7
+
+    #: A standard Python attribute or descriptors (no validation or
```
Comment (mdickinson) [inline traits/constants.py]: typo: descriptor

File: traits/api.py [hunk 4]
```diff
@@ -230,3 +229,8 @@
...
+NO_COMPARE = ComparisonMode.no_compare
```
Comment (mdickinson) [inline traits/api.py]: Suggest adding `#:`-style docstrings here. Then we can document these constants in the API documentation (at least to the extent that a search for `OBJECT_IDENTITY_COMPARE` brings up a result).

File: traits/constants.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    float = 21
+
+
+class ComparisonMode(IntEnum):
+    """ Comparison mode.
+
+    Indicates when trait change notifications should be generated based upon
+    the result of comparing the old and new values of a trait assignment:
+
+    * 0 (no_compare): The values are not compared and a trait change
```
Comment (mdickinson) [inline traits/constants.py]: This seems redundant with the `#:`-style docstrings below, but I couldn't see a way to persuade Sphinx to render the individual docstrings. :-(

Comment (mdickinson) [reply traits/constants.py]: Related: https://github.com/sphinx-doc/sphinx/issues/5679

**Comment by mdickinson** [comment]
Will fix merge conflicts shortly.

**Comment by mdickinson** [comment]
Assigning to me for merge conflict resolution, minor updates, and merge.

File: traits/api.py (hunk 4 above)
Comment (mdickinson) [reply traits/api.py]: I'll do this in a separate PR.

File: traits/constants.py (hunk 5 above)
Comment (mdickinson) [reply traits/constants.py]: Nothing to do here. Resolving.

File: traits/api.py (hunk 1 above)
Comment (corranwebster) [reply traits/api.py]: Yes, I think it makes sense to encourage the use of the Enums.

Comment (mdickinson) [reply traits/api.py]: Okay. I added a note to that effect to #695.

**Pull request merged**